### PR TITLE
Debugger Client print feature

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

File: src/shell/Shell.cpp [hunk 1]
```diff
@@ -77,8 +77,10 @@ ValueRef* builtinPrint(ExecutionStateRef* state, ValueRef* thisValue, size_t arg
     if (argc >= 1) {
         if (argv[0]->isSymbol()) {
             puts(argv[0]->asSymbol()->symbolDescriptiveString()->toStdUTF8String().data());
+            state->context()->printDebugger(argv[0]->asSymbol()->symbolDescriptiveString()->toStdUTF8String().data());
         } else {
             puts(argv[0]->toString(state)->toStdUTF8String().data());
+            state->context()->printDebugger(argv[0]->toString(state)->toStdUTF8String().data());
```
Comment (zherczeg) [inline src/shell/Shell.cpp:83]: Since we send strings here, the API should accept Escargot string, rather than char*.

File: tools/debugger/debugger_core.py [hunk 2]
```diff
@@ -850,9 +860,9 @@ def _receive_string(self, message_type, prefix, data):
         result += data[1:]
 
         if end_type == message_type + 1:
-            result = self.encode8(result)
+            result = self.dencode8(result)
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:863]: I think this will be fixed in another patch. It will be decode, not dencode :)

File: src/debugger/Debugger.cpp [hunk 3]
```diff
@@ -437,6 +437,15 @@ void Debugger::getScopeChain(ExecutionState* state)
     send(ESCARGOT_MESSAGE_SCOPE_CHAIN_END, buffer, nextScope);
 }
 
+void Debugger::sendPrint(const char* output)
+{
+    StringBuilder builder;
+    builder.appendString("Print: ");
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply src/debugger/Debugger.cpp:443]: I mean adding the "Print: " prefix there. It is easer to do that in python/js side.

**Comment by clover2123** [comment]
Please rebase this patch.

File: src/shell/Shell.cpp [hunk 4]
```diff
@@ -77,8 +77,10 @@ ValueRef* builtinPrint(ExecutionStateRef* state, ValueRef* thisValue, size_t arg
     if (argc >= 1) {
         if (argv[0]->isSymbol()) {
             puts(argv[0]->asSymbol()->symbolDescriptiveString()->toStdUTF8String().data());
+            state->context()->printDebugger(argv[0]->asSymbol()->symbolDescriptiveString());
```
Comment (zherczeg) [inline src/shell/Shell.cpp:80]: You could put `argv[0]->asSymbol()->symbolDescriptiveString()` into a local variable.

File: src/shell/Shell.cpp [hunk 5]
```diff
@@ -77,8 +77,10 @@ ValueRef* builtinPrint(ExecutionStateRef* state, ValueRef* thisValue, size_t arg
...
         } else {
             puts(argv[0]->toString(state)->toStdUTF8String().data());
+            state->context()->printDebugger(argv[0]->toString(state));
```
Comment (zherczeg) [inline src/shell/Shell.cpp:83]: Ditto.

File: src/api/EscargotPublic.cpp [hunk 6]
```diff
@@ -1626,6 +1626,16 @@ bool ContextRef::initDebugger(const char* options)
 #endif /* ESCARGOT_DEBUGGER */
 }
 
+void ContextRef::printDebugger(StringRef* output)
+{
+#ifdef ESCARGOT_DEBUGGER
+    StringView* outputStr = new StringView(toImpl(output));
+    toImpl(this)->printDebugger(outputStr);
+#else /* !ESCARGOT_DEBUGGER */
+    return;
```
Comment (zherczeg) [inline src/api/EscargotPublic.cpp:1635]: No need return here.
